src/Stratis.Bitcoin.Tests/Utilities/ParallelAsyncTest.cs

```diff
@@ -29,36 +29,33 @@ public ParallelAsyncTest()
         {
             int sum = 0;
 
-            Stopwatch watch = Stopwatch.StartNew();
+            var delayTask = new Task(async () => { await Task.Delay(this.testCollection.Count * this.itemProcessingDelayMs / 2).ConfigureAwait(false); });
 
             await this.testCollection.ForEachAsync(2, CancellationToken.None, async (item, cancellation) =>
             {
-                await Task.Delay(this.itemProcessingDelayMs).ConfigureAwait(false);
                 sum += item;
```

Comment:
You've got race condition here

Reply:
That's basically the intention: create a race condition where we expect something to finish first instead of using time measurements with StopWatch

Reply:
Unlikely, the test will fail if the race condition happens, you are asserting that the sum you are computing will equal to the sum of all items

Reply:
There is nothing wrong with the stopwatch, the whole test is bogus and race condition is just an extra reason like fail

the test will fail quite likely the machine with single processor and single thread execution and on multi-core machine the sample should be much larger in order for the test to have only very low chance of failing

Reply:
How is that possible if the sum is updated before the awaiting of the Task.Delay? There should be no way we execute `Assert.Equal(this.testCollectionSum, sum);` before we summarize all items.

Reply:
Though the race condition can happen in ForEachAsync_CanBeCancelled_Async, that I'll fix 🙂

Reply:
`sum += item` Can be executed by 2 threads in parallel, at least I believe that is the case, is it?

Reply:
Although the delay in the task is too low, so that's just another point where this test can fail

Reply:
>sum += item Can be executed by 2 threads in parallel, at least I believe that is the case, is it?

correct, it can.

Reply:
Still don't see where the race condition can happen which will lead to execution `Assert.Equal(this.testCollectionSum, sum);` before all the items from the list were summarized. 

Maybe you are talking about prev version where there was delay first and after that sum update? If that's the case- yes, race here can happen and that is fixed in the 1st commit of that PR.

Reply:
`sum += item` can be skipped due to race condition

Reply:
Fixed